RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc

```diff
@@ -1,425 +1,475 @@
+				const Propagator& propagatorAlong,
+				const Propagator& propagatorOpposite,
+				const reco::TrackRef l2,
+				edm::ESHandle<Chi2MeasurementEstimatorBase>& estimatorH,
+				edm::Handle<MeasurementTrackerEvent>& measurementTrackerH,
+				unsigned int& numSeedsMade,
+                unsigned int& hitlessSeedsMade,
+				unsigned int& numOfMaxSeeds,
+				unsigned int& layerCount,
+				std::unique_ptr<std::vector<TrajectorySeed> >& out) const {
+  
+  if (numSeedsMade > numOfMaxSeeds) return;
+  LogTrace("TSGForOI") << "TSGForOI::findSeedsOnLayer: numSeedsMade = " << numSeedsMade << " , layerCount = " <<  layerCount << endl;
+  
+  double errorSFHits=1.0;
+  double errorSFHitless=1.0;
+  if (!adjustErrorsDynamicallyForHits_) errorSFHits = fixedErrorRescalingForHits_;
+  else                                  errorSFHits = calculateSFFromL2(l2);
+  if (!adjustErrorsDynamicallyForHitless_) errorSFHitless = fixedErrorRescalingForHitless_;
+  else                                  errorSFHitless = calculateSFFromL2(l2);
+
+  // create hitless seeds
+  // create one hitless seed per L2 on  on each layer
+  if ( useHitLessSeeds_ ) {
+    LogTrace("TSGForOI") << "TSGForOI::findSeedsOnLayer: Start hitless" << endl;
+    std::vector< GeometricSearchDet::DetWithState > dets;
+    layer.compatibleDetsV(tsosAtIP, propagatorAlong, *estimatorH, dets);
+    if (!dets.empty()) {
+      auto const& detOnLayer = dets.front().first;
+      auto const& tsosOnLayer = dets.front().second;
+      LogTrace("TSGForOI") << "TSGForOI::findSeedsOnLayer: tsosOnLayer " << tsosOnLayer << endl;
+      if (!tsosOnLayer.isValid()) {
+        edm::LogInfo(theCategory) << "ERROR!: Hitless TSOS is not valid!";
+      }
+      else {
+        dets.front().second.rescaleError(errorSFHitless);
+        PTrajectoryStateOnDet const& ptsod = trajectoryStateTransform::persistentState(tsosOnLayer,detOnLayer->geographicalId().rawId());
+        TrajectorySeed::recHitContainer rHC;
+        out->push_back(TrajectorySeed(ptsod,rHC,oppositeToMomentum));
+        LogTrace("TSGForOI") << "TSGForOI::findSeedsOnLayer: TSOD (Hitless) done " << endl;
+        hitlessSeedsMade++;
+      }
+    }
+  }
+
+  // Hits
+  if (layerCount > numOfLayersToTry_) return;
+  LogTrace("TSGForOI") << "TSGForOI::findSeedsOnLayer: Start Hits" << endl;
+  if (makeSeedsFromHits(tTopo, layer, tsosAtIP, *out, propagatorAlong, *measurementTrackerH, estimatorH, numSeedsMade, errorSFHits, l2->eta())) ++layerCount;
+
+}
+
+
+//
+// Calculate the dynamic error SF by analysing the L2
+//
+double TSGForOI::calculateSFFromL2(const reco::TrackRef track) const {
+
+  double theSF = 1.0;
+  //	L2 direction vs pT blowup - as was previously done:
+  //	Split into 4 pT ranges: <pT1_, pT1_<pT2_, pT2_<pT3_, <pT4_: 13,30,70
+  //	Split into different eta ranges depending in pT
+
+
+  double abseta = std::abs(track->eta());
+  if (track->pt() <= pT1_) theSF = SF1_;
+  if (track->pt() > pT1_ && track->pt() <= pT2_) {

```

Comment:
Assuming that all ```pt[i]_``` and ```eta[i]_``` are sorted, here and below all mutually excluded pt/eta ranges should be checked with ```else if```